### PR TITLE
Late Linking WIP

### LINKED ISSUE #70: Improve libraries to be more usable.
Right now the extent of `"libraries"` is to add them to the genesis block, and supply a fixed address to `solc` when compiling contracts, as we dont support unlinked bytecode.

We'd like the ability to:
1. Specify different addresses on different networks for the same library, so that fetching the bytecode is much more flexible. It's highly unlikely the same library will have the same address across the mainnet and testnets.
2. Deploy libraries as part of our project, whether these are libraries defined in a dependency that we have source available for or a library defined in our own project.
3. Stop feeding the library addresses to `solc` during compile and dynamically link in the library when deploying depending on what network we're connecting to. This means we need to support unlinked bytecode and inject the `___LIB_NAME________` addresses as the deploy happens. In principle, this also means we'd want to have different artifact sets based on the network we're using.

### PR DESCRIPTION
**Pull request by iostat**
This is a WIP/tracking PR to essentially fix #70. There's still a bunch of ergonomics stuff to be done, mainly around how linked bytecode is stored and used (to prevent things like contracts linked against libraries on network X being deployed on network Y, etc.). Additionally, doing this correctly makes #85 much easier going forward.

Please don't merge this yet.

### CODE REVIEW

File: src/Chanterelle/Internal/Utils/Error.purs [hunk 1]
```diff
@@ -0,0 +1,102 @@
+module Chanterelle.Internal.Utils.Error where
```
Comment (safareli) [inline src/Chanterelle/Internal/Utils/Error.purs]: Some of this functions are not used and it seams like some of them don't quite pass the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold).

File: src/Chanterelle/Project.purs [hunk 2]
```diff
@@ -37,17 +70,29 @@ loadProject root = do
...
+      libJsonOut = Path.concat [root, project.libArtifactsDir]
+      psOut      = Path.concat [root, project.psGen.outputPath]
+      srcIn      = Path.concat [root, project.sourceDir]
+      modules    = mkModule <$> project.modules
+      libModules = mkLibModule <$> libs
       mkModule moduleName =
         let solPath      = Path.concat [srcIn, pathModName <> ".sol"]
             jsonPath     = Path.concat [jsonOut, pathModName <> ".json"]
             pursPath     = Path.concat [psOut, psModBase, pathModName <> ".purs"]
-            solContractName = unsafePartialBecause "String.split always returns a non-empty Array" $
```
Comment (safareli) [inline src/Chanterelle/Project.purs]: for fun:
```js
> "a".split("a")
Array [ "", "" ]

> "".split("")
Array []
```

File: src/Chanterelle/Internal/Types/Project.purs [hunk 3]
```diff
@@ -215,8 +204,8 @@ instance encodeJsonSolcOptimizerSettings :: EncodeJson SolcOptimizerSettings whe
 instance decodeJsonSolcOptimizerSettings :: DecodeJson SolcOptimizerSettings where
   decodeJson j = do
       obj <- decodeJson j
-      enabled <- fromMaybe default.enabled <$> obj .?? "enabled"
```
Comment (safareli) [inline src/Chanterelle/Internal/Types/Project.purs]: FYI there is sequenceRecord in Record.Extra which can be used like this:
```purs
Foo <$> sequenceRecord
  { a: obj .? "a"
  , b: obj .? "b"
  }
```

File: src/Chanterelle/Internal/Utils/Error.purs (hunk 1 above)
Comment (blinky3713) [reply src/Chanterelle/Internal/Utils/Error.purs]: I would also agree with this

**Comment by iostat** [comment]
Yeah a bunch of those error helpers I made mid-refactor to help preserve old behavior in certain places while making the flow more readable. Definitely agree that most of them should go as  half of them ended up being unused by the end.

**Pull request merged**